### PR TITLE
kubelet/cm: bump runc/libcontainer to 1.1.0

### PR DESCRIPTION
**Pull request by kolyshkin**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

TL;DR: simplify/improve cgroup management code in kubelet/cm. Bump to runc 1.1.0.

This updates vendored runc/libcontainer to 1.1.0
(release notes at https://github.com/opencontainers/runc/releases)
and google/cadvisor to a version with PR https://github.com/google/cadvisor/pull/3031 merged.

Changes in vendor are generated by (roughly):
```
  ./hack/pin-dependency.sh github.com/google/cadvisor v0.44.0
  ./hack/pin-dependency.sh github.com/opencontainers/runc v1.1.0
  ./hack/update-vendor.sh
  ./hack/lint-dependencies.sh # And follow all its recommendations.
  ./hack/update-vendor.sh
  ./hack/update-internal-modules.sh
  ./hack/lint-dependencies.sh # Re-check everything again.
```

Changes in pkg/kubelet/cm are to adopt new runc 1.1 libcontainer/cgroups
APIs and build on its improvements. In particular:
    
 - simplify cgroup manager instantiation, using a new, easier way of
       `libcontainers/cgroups/manager.New`;

 - replace `libcontainerAdapter` with a boolean variable (all it did
       was passing on whether systemd manager should be used);
    
 - trivial change due to removed `cgroupfs.HugePageSizes` and added
        `cgroups.HugePageSizes()`;
    
 - do not calculate cgroup paths in update / destroy, since libcontainer
       cgroup managers now calculate the paths upon creation (previously,
       they were doing that only in `Apply`, so using e.g. `Set` or `Destroy` right
       after creation was impossible without specifying paths).
       🔔 We currently still calculate cgroup paths in `Exists` -- this is to be
        addressed separately.

#### Which issue(s) this PR fixes:
none

#### Special notes for your reviewer:

* Please review commit by commit, and see individual commit messages for more details.

* This is currently a draft, as the code is on top of #108597. Once that PR is merged, this will be rebased.

#### Does this PR introduce a user-facing change?
```release-note
Switch to opencontainers/runc/libcontainer v1.1.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
none

### CODE REVIEW

**Comment by kolyshkin** [comment]
OK, it works with these 4 commits. Let's add more.

**Comment by kolyshkin** [comment]
The `TestCronJobLaunchesPodAndCleansUp` failure in  `pull-kubernetes-integration` seems to be an unrelated (but recent) flake, first observed in https://github.com/kubernetes/kubernetes/pull/106909#issuecomment-998553331

/test pull-kubernetes-integration

**Comment by kolyshkin** [comment]
Updated to runc-1.1.0 final.

**Comment by kolyshkin** [comment]
Rebased.

/remove-sig api-machinery
/remove-sig instrumentation

**Comment by kolyshkin** [comment]
@odinuge PTAL

**Comment by kolyshkin** [comment]
Rebased after #105142 merge.

/remove-sig api-machinery
/remove-sig instrumentation

**Comment by mrunalp** [comment]
@giuseppe ptal

**Review by giuseppe — APPROVED** [review]
LGTM

**Comment by ehashman** [comment]
/priority important-soon
/triage accepted

We'll need to bump this in this release assuming we bump cadvisor, which I believe we will. But I think we want to bump runc in cadvisor first, and that doesn't appear to have landed in a release yet.

**Comment by bobbypage** [comment]
new tagged version of cAdvisor has been released with runc 1.1.0 so it can be used to unblock this PR - https://github.com/google/cadvisor/releases/tag/v0.44.0

**Comment by kolyshkin** [comment]
Updated to cAdvisor v0.44.0; this is ready for review and merge.

File: pkg/kubelet/cm/cgroup_manager_linux.go [hunk 1]
```diff
@@ -180,42 +139,39 @@ type CgroupSubsystems struct {
...
 // Make sure that cgroupManagerImpl implements the CgroupManager interface
 var _ CgroupManager = &cgroupManagerImpl{}
 
 // NewCgroupManager is a factory method that returns a CgroupManager
-func NewCgroupManager(cs *CgroupSubsystems, cgroupDriver string) CgroupManager {
-	managerType := libcontainerCgroupfs
-	if cgroupDriver == string(libcontainerSystemd) {
-		managerType = libcontainerSystemd
-	}
+func NewCgroupManager(cs *CgroupSubsystems, useSystemd bool) CgroupManager {
```
Comment (ehashman) [inline pkg/kubelet/cm/cgroup_manager_linux.go:156]: I think it's a bit confusing to change the type signature of this function here... while the compiler is going to ensure we don't miss any conversions, I think this might be confusing when trying to compare old vs. new code. If it's possible to lift this out of sight and avoid the change I'd do that, e.g. 

```suggestion
func NewCgroupManager(cs *CgroupSubsystems, cgroupDriver string) CgroupManager {
    useSystemd := cgroupDriver == "systemd"
```

We may also want to factor out the constant for the systemd cgroup driver name.

Comment (kolyshkin) [reply pkg/kubelet/cm/cgroup_manager_linux.go:156]: I have changed that because the old interface allows an arbitrary string and won't catch the error of providing some arbitrary value.

Changed back to string.

**Comment by kolyshkin** [comment]
I'm not sure what to make of test failures -- are all of them currently broken?

**Comment by pacoxu** [comment]
> I'm not sure what to make of test failures -- are all of them currently broken?

I think it is related to https://github.com/kubernetes/test-infra/issues/24798

File: pkg/kubelet/cm/node_container_manager_linux.go [hunk 2]
```diff
@@ -152,13 +155,7 @@ func enforceExistingCgroup(cgroupManager CgroupManager, cName CgroupName, rl v1.
 		Name:               cName,
 		ResourceParameters: rp,
 	}
-	if cgroupConfig.ResourceParameters == nil {
-		return fmt.Errorf("%q cgroup is not config properly", cgroupConfig.Name)
-	}
 	klog.V(4).InfoS("Enforcing limits on cgroup", "cgroupName", cName, "cpuShares", cgroupConfig.ResourceParameters.CpuShares, "memory", cgroupConfig.ResourceParameters.Memory, "pidsLimit", cgroupConfig.ResourceParameters.PidsLimit)
-	if !cgroupManager.Exists(cgroupConfig.Name) {
```
Comment (odinuge) [inline pkg/kubelet/cm/node_container_manager_linux.go:159]: Hmm. Not 100% sure about this one. The `cgroupManager.Exists` will emit all the missing control groups / controllers, while I assume the runc `Set` will not, and fail on the first one. Not sure if it important tho., but I assume we can take it in a separate PR and take a deep dive.

File: cmd/kubelet/app/server.go [hunk 3]
```diff
@@ -606,7 +606,7 @@ func run(ctx context.Context, s *options.KubeletServer, kubeDeps *kubelet.Depend
 	}
 
 	var cgroupRoots []string
-	nodeAllocatableRoot := cm.NodeAllocatableRoot(s.CgroupRoot, s.CgroupsPerQOS, s.CgroupDriver)
+	nodeAllocatableRoot := cm.NodeAllocatableRoot(s.CgroupRoot, s.CgroupsPerQOS, s.CgroupDriver == "systemd")
```
Comment (odinuge) [inline cmd/kubelet/app/server.go:609]: Can we keep the old cgroup driver signature here as well, and then do the refactor in another one as suggested by @ehashman (or did you just suggest for a single instance)?

**Review by odinuge — COMMENTED** [review]
Took an initial look, and it looks ok at first glance. The runc changes are so big this time, so haven't had a deep dive into them, but I think @kolyshkin have more control on that front than me. I am leaning towards splitting the other non-bump-related changes to separate PR(s), but I think a lot of them are valuable. Feedback from other reviewers would be appreciated. I do have a few questions tho.

The CI situation is kinda sad, and I would prefer if it was sorted before merging this. The failing suites are the one testing cgroup v2 and systemd support, and having them working is kinda important. Do you know if someone are working actively on fixing it @pacoxu?

File: pkg/kubelet/cm/cgroup_manager_linux.go [hunk 4]
```diff
@@ -89,7 +89,7 @@ func unescapeSystemdCgroupName(part string) string {
 // This function always expands the systemd name into the cgroupfs form. If only
 // the last part is needed, use path.Base(...) on it to discard the rest.
 func (cgroupName CgroupName) ToSystemd() string {
-	if len(cgroupName) == 0 || (len(cgroupName) == 1 && cgroupName[0] == "") {
+	if len(cgroupName) == 0 {
```
[the remark this answers is not in the document]
Comment (kolyshkin) [reply pkg/kubelet/cm/cgroup_manager_linux.go:84]: 🤦🏻 

For some reason I read the code thinking that `CgroupName` is `string` rather than `[]string`. It surely makes sense with `[]string`. 

Removed this commit; thanks for catching this.

File: pkg/kubelet/cm/cgroup_manager_linux.go [hunk 5]
```diff
@@ -534,7 +534,7 @@ func (m *cgroupManagerImpl) Create(cgroupConfig *CgroupConfig) error {
 
 	// it may confuse why we call set after we do apply, but the issue is that runc
 	// follows a similar pattern.  it's needed to ensure cpu quota is set properly.
-	if err := m.Update(cgroupConfig); err != nil {
+	if err := manager.Set(resources); err != nil {
```
[the remark this answers is not in the document]
Comment (kolyshkin) [reply pkg/kubelet/cm/cgroup_manager_linux.go]: This is a separate commit (I guess you are not reviewing this commit-by-commit, since this is the first commit).

I can certainly move it out to a separate PR and made this one a draft waiting for that other PR to be merged. Same for other 5 preparatory commits.

File: pkg/kubelet/cm/cgroup_manager_linux.go [hunk 6]
```diff
@@ -449,8 +449,7 @@ func (m *cgroupManagerImpl) Update(cgroupConfig *CgroupConfig) error {
 	}()
 
 	// Extract the cgroup resource parameters
-	resourceConfig := cgroupConfig.ResourceParameters
-	resources := m.toResources(resourceConfig)
+	resources := m.toResources(cgroupConfig.ResourceParameters)
```
[the remark this answers is not in the document]
Comment (kolyshkin) [reply pkg/kubelet/cm/cgroup_manager_linux.go]: Again, this is a separate commit.

I'm not very familiar with kubernetes processes. Isn't it possible to revert separate commits, rather than the whole PR?

**Comment by kolyshkin** [comment]
This PR is 6 commits, of which 5 are preparatory/cleanup, and the last one is the actual runc bump. I guess it makes more sense if you can review it commit by commit.

I can move these preparatory commits out of this PR (and make this one a draft until those commits are merged, and then do a rebase and mark it as ready for review).

@odinuge Would you prefer a separate PR for each of the 5 prep commits, or a single PR for all of them?

File: cmd/kubelet/app/server.go (hunk 3 above)
Comment (kolyshkin) [reply cmd/kubelet/app/server.go]: Done.

File: pkg/kubelet/cm/node_container_manager_linux.go (hunk 2 above)
Comment (kolyshkin) [reply pkg/kubelet/cm/node_container_manager_linux.go:159]: Right, `cgroupManager.Exists` will report all missing cgroups on error, while a runc cgroup manager will only report the first missing cgroup. I'm not sure if there is a difference practically, since
 - normally they all either exist or not;
 - this code is assuming they all exist.
 
 If there is a need for detailed info, this call can be moved to error path (i.e. call `Exists` when `Update` failed).
 
 I'll rework it and move out of here.

Comment (kolyshkin) [reply pkg/kubelet/cm/node_container_manager_linux.go:159]: Brought the Exists check back.

**Comment by kolyshkin** [comment]
Rebased.

**Comment by kolyshkin** [comment]
> @odinuge Would you prefer a separate PR for each of the 5 prep commits, or a single PR for all of them?

Did a single PR: https://github.com/kubernetes/kubernetes/pull/108597; PTAL

**Comment by kolyshkin** [comment]
Rebased.

File: pkg/kubelet/cm/cgroup_manager_linux.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (odinuge) [reply pkg/kubelet/cm/cgroup_manager_linux.go]: > This is a separate commit (I guess you are not reviewing this commit-by-commit, since this is the first commit).

Yeah, thanks for doing that! I definitely prefer your style of doing it with separate commits with well defined commit messages! I am a huge fan of that, thanks!

 > I can certainly move it out to a separate PR and made this one a draft waiting for that other PR to be merged. Same for other 5 preparatory commits.
 
 Thanks. The only worry is doing a lot of these cleanups in the same PR as a big dependency change; mostly for being (better) able to bisect the CI running on the master branch, as well as the "work" of reverting if we have to do that. Other than for dependency changes or verrry big PRs, I do prefer your style waay over the alternative. Again, thanks for doing this!

**Comment by odinuge** [comment]
> Did a single PR: https://github.com/kubernetes/kubernetes/pull/108597; PTAL

Thanks @kolyshkin! Reviewed that one separately. Feel free to hide (or what it is called on github) my earlier PR reviews, since they no longer apply to this PR.

I am still a bit worried about https://github.com/kubernetes/test-infra/issues/24798 still being broken, meaning we have no CI signal for cgroup v2 (I still don't think we have cgroup v2 containerd tests, but I might be wrong..), the systemd cgroup driver or cri-o...

**Comment by kolyshkin** [comment]
Rebased on top of just-merged #108597; no longer a draft

**Review by odinuge — APPROVED** [review]
Changes looks good to me now! I have looked through the code with the previous regressions we have seen, and I haven't found any red flags this time. Thanks for pushing @kolyshkin 

/lgtm

But lets hold _some_ time to let the CI run for the other changes (eg. this run https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-cos-cgroupv2-containerd-node-e2e-serial/1506700583120146432 of `ci-cos-cgroupv2-containerd-node-e2e-serial`).

Feel free to unhold when we have CI confidence about the other change. 
/hold

**Comment by odinuge** [comment]
And lets test this one that failed; 

/test pull-kubernetes-node-swap-ubuntu-serial

As well as this one that should pass;

/test pull-kubernetes-node-kubelet-serial-containerd

**Comment by ehashman** [comment]
/milestone v1.24

@kolyshkin can you please remove draft status on the PR and s/Switch/Update/ in the change note?

**Comment by ehashman** [comment]
I think this needs rebase...

**Comment by bobbypage** [comment]
This has been replaced with https://github.com/kubernetes/kubernetes/pull/109029
